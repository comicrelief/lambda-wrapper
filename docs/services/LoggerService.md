# LoggerService

Provides logging and integrations with our monitoring tools.

For logging we use [Winston](https://github.com/winstonjs/winston). Errors will also be sent to [Sentry](https://sentry.io/) and [Epsagon](https://epsagon.com/) if those are configured.

## Usage

The logger exposes various methods that you can pass messages or objects to for logging:

```ts
import lambdaWrapper, { LoggerService } from '@comicrelief/lambda-wrapper';

export default lambdaWrapper.wrap(async (di) => {
  const logger = di.get(LoggerService);

  // general log message
  logger.info('Doing something');

  // tag the trace so we can find certain tracess more easily in Epsagon
  logger.label('flag');
  logger.metric('transactionId', value);

  try {
    // do something that might throw an error...
  } catch (error) {
    // log the error and flag the trace on Epsagon and Sentry
    logger.error(error);

    // alternatively, use `warning` if this error is not relevant in staging
    // (see Soft Warnings below)
    logger.warning(error);
  }
});
```

## Configuration

### Soft warnings

The `warning` method is equivalent to `error` by default, but can be switched to use `info` by setting `LOGGER_SOFT_WARNING=1` in the environment.

This is handy for muting certain errors in staging, where we expect our integration tests to cause a lot of errors deliberately that would otherwise spam us with Epsagon alerts.

### Epsagon

To configure Epsagon, set the following environment variables:

- `EPSAGON_TOKEN` – your access token
- `EPSAGON_SERVICE_NAME` – the application name (including stage) to record traces under

### Sentry

To configure Sentry, set the following environment variables:

- `RAVEN_DSN` – your Sentry DSN URL
