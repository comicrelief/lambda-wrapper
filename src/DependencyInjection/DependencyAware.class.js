/**
 * DependencyAwareClass Class
 */
export default class DependencyAwareClass {
  /**
   * DependencyAwareClass constructor
   * @param di DependencyInjection
   */
  constructor(di) {
    this.di = di;
  }

  /**
   * Get Dependency Injection Container
   * @return DependencyInjection
   */
  getContainer() {
    return this.di;
  }
}
