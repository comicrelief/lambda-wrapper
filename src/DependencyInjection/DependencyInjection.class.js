import { DEFINITIONS, DEPENDENCIES } from '../Config/Dependencies';

/**
 * DependencyInjection class
 */
export default class DependencyInjection {
  /**
   * DependencyInjection constructor
   * @param configuration
   * @param event
   * @param context
   */
  constructor(configuration, event, context) {
    this.event = event;
    this.context = context;

    this.dependencies = {};
    this.configuration = configuration;

    for (let x = 0; x <= 1; x += 1) {
      // Iterate over lapper dependencies and add to container
      Object.keys(DEFINITIONS).forEach((dependencyKey) => {
        this.dependencies[dependencyKey] = new DEPENDENCIES[dependencyKey](this);
      });

      // Iterate over child dependencies and add to container
      if (typeof configuration.DEPENDENCIES !== 'undefined') {
        Object.keys(configuration.DEPENDENCIES).forEach((dependencyKey) => {
          this.dependencies[dependencyKey] = new configuration.DEPENDENCIES[dependencyKey](this);
        });
      }
    }
  }

  /**
   * Get Dependency
   * @param definition
   * @return {*}
   */
  get(definition) {
    if (typeof this.dependencies[definition] === 'undefined') {
      throw Error(`${definition} does not exist in di container`);
    }

    return this.dependencies[definition];
  }

  /**
   * Get Event
   * @return {*}
   */
  getEvent() {
    return this.event;
  }

  /**
   * Get Context
   * @return {*}
   */
  getContext() {
    return this.context;
  }

  /**
   * Get Configuration
   * @param definition string
   * @return {*}
   */
  getConfiguration(definition = null) {
    if (definition !== null && typeof this.configuration[definition] === 'undefined') {
      return null;
    } else if (typeof this.configuration[definition] !== 'undefined') {
      return this.configuration[definition];
    }

    return this.configuration;
  }
}
