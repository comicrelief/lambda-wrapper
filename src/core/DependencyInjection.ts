import { Context } from 'aws-lambda';

import DependencyAwareClass from './DependencyAwareClass';
import { LambdaWrapperConfig } from './config';

// eslint-disable-next-line no-use-before-define
type Class<TInstance, TConfig extends LambdaWrapperConfig> = new (di: DependencyInjection<TConfig>) => TInstance;

/**
 * Dependency injection container.
 *
 * Dependencies (singleton instances of dependency-aware classes) are provided
 * to the main Lambda handler and other dependencies via this class.
 */
export default class DependencyInjection<TConfig extends LambdaWrapperConfig = any> {
  /**
   * Instantiated dependencies.
   */
  readonly dependencies: Record<string, DependencyAwareClass>;

  /**
   * True until all dependencies have been constructed.
   */
  private isConstructing = true;

  constructor(
    readonly config: TConfig,
    readonly event: any,
    readonly context: Context,
  ) {
    // get unique dependency classes -- a class may be included several times,
    // but should be instantiated only once
    const classes = Array.from(new Set(Object.values(config.dependencies)));

    // guard against duplicate keys
    const countByName = classes
      .map((Constructor) => Constructor.name)
      .reduce(
        (counts, name) => ({ ...counts, [name]: (counts[name] || 0) + 1 }),
        {} as Record<string, number>,
      );
    if (Object.values(countByName).some((count) => count > 1)) {
      const duplicateNames = Object.entries(countByName)
        .filter(([, count]) => count > 1)
        .map(([name]) => name);

      // if all class names are single-letter, they're probably minified -- in
      // this case, give a hint about how to fix it
      const action = duplicateNames.every((it) => it.length === 1)
        ? "If you don't recognise the single-letter names listed above, your "
          + "bundler may be minifying your code. You'll need to disable this "
          + 'for Lambda Wrapper to work correctly. Please refer to the Notes '
          + 'section of the Lambda Wrapper readme:\n\n'
          + '  https://github.com/comicrelief/lambda-wrapper#notes'
        : 'Please ensure that all dependency classes have a unique name.';

      throw new Error(
        `Dependency names are not unique: ${duplicateNames.join(', ')}\n\n${action}`,
      );
    }

    // instantiate all dependencies
    this.dependencies = Object.fromEntries(
      classes.map((Constructor) => [Constructor.name, new Constructor(this)]),
    );

    this.isConstructing = false;
  }

  /**
   * Get the singleton instance of the given dependency.
   *
   * @param dependency
   */
  get<T extends DependencyAwareClass>(dependency: Class<T, TConfig>): T {
    if (this.isConstructing) {
      throw new Error(
        'Dependencies are not available in dependency class constructors.\n\n'
        + 'To fix this, call `di.get` in the function where the dependency is '
        + 'used instead of inside your constructor.',
      );
    }

    const name = dependency.name;

    if (!this.dependencies[name]) {
      throw new Error(
        `${name} does not exist in dependency container\n\n`
        + `Make sure you've included ${name} in the 'dependencies' key of your `
        + 'Lambda Wrapper config.',
      );
    }

    return this.dependencies[name] as T;
  }

  /**
   * Get the event passed to AWS Lambda.
   *
   * @deprecated Use `di.event` instead.
   */
  getEvent() {
    return this.event;
  }

  /**
   * Get the AWS Lambda context object.
   *
   * @deprecated Use `di.context` instead.
   */
  getContext() {
    return this.context;
  }

  /**
   * Get Lambda Wrapper configuration.
   *
   * @deprecated Use `di.config` instead.
   */
  getConfiguration() {
    return this.config;
  }

  /**
   * True if the function is being executed in `serverless-offline`.
   *
   * We use the following checks for this:
   *
   * - if there is no function ARN, or the ARN includes 'offline'
   * - if `process.env.USE_SERVERLESS_OFFLINE` is set
   *
   * TODO: This is nothing to do with dependency injection and should be moved
   *   somewhere else! Any ideas?
   */
  get isOffline(): boolean {
    return !this.context.invokedFunctionArn
      || this.context.invokedFunctionArn.includes('offline')
      || !!process.env.USE_SERVERLESS_OFFLINE;
  }

  /**
   * Get the `service-stage-` prefix added by Serverless to deployed Lambda
   * function names. This is handy when you want to invoke other functions,
   * without having to hardcode the service name and stage.
   *
   * The returned prefix includes a trailing dash. To get the deployed name of
   * another Lambda function, concatenate its serverless function name (its key
   * in `serverless.yml`) onto the prefix:
   *
   * ```js
   * const serverlessFunctionName = 'MyFunction';
   * const deployedName = `${di.getLambdaPrefix()}${serverlessFunctionName}`;
   * ```
   *
   * This function relies on looking at the currently running Lambda function's
   * resource name. It will not work correctly if the Lambda function has been
   * given a custom resource name.
   */
  getLambdaPrefix(): string {
    const stage = process.env.STAGE;
    if (!stage) {
      /* eslint-disable no-template-curly-in-string */
      throw new Error(
        'STAGE is not set\n\n'
        + 'Please add to your Lambda environment:\n\n'
        + '    STAGE: ${sls:stage}\n',
      );
    }
    if (!this.context.functionName) {
      throw new Error('Lambda function name is unavailable in context');
    }
    const [service] = this.context.functionName.split(`-${stage}-`);
    return `${service}-${stage}-`;
  }
}
