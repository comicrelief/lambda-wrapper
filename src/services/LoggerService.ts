import * as lumigo from '@lumigo/tracer';
import * as Sentry from '@sentry/node';
import { AxiosError } from 'axios';
import Winston from 'winston';

import DependencyAwareClass from '../core/DependencyAwareClass';
import DependencyInjection from '../core/DependencyInjection';

const sentryIsAvailable = typeof process.env.RAVEN_DSN !== 'undefined' && typeof process.env.RAVEN_DSN === 'string' && process.env.RAVEN_DSN !== 'undefined';

// initialise the Sentry client if available
if (sentryIsAvailable) {
  Sentry.init({
    dsn: process.env.RAVEN_DSN,
    shutdownTimeout: 5,
    environment: process.env.STAGE,
  });
}

/**
 * Provides logging and integrations with our monitoring tools.
 *
 * For logging we use [Winston](https://github.com/winstonjs/winston).
 * Errors will also be sent to [Sentry](https://sentry.io/) and
 * [Epsagon](https://epsagon.com/) if those are available.
 */
export default class LoggerService extends DependencyAwareClass {
  private sentry: typeof Sentry | null;

  private winston: Winston.Logger | null;

  constructor(di: DependencyInjection) {
    super(di);

    this.sentry = null;
    this.winston = null;

    const { event, context } = this.di;

    if (sentryIsAvailable && !di.isOffline) {
      Sentry.configureScope((scope) => {
        scope.setTags({
          Event: event,
          Context: context as any,
        });
        scope.setExtras({
          lambda: context.functionName,
          memory_size: context.memoryLimitInMB,
          log_group: context.logGroupName,
          log_stream: context.logStreamName,
          stage: process.env.STAGE,
          path: event.path,
          httpMethod: event.httpMethod,
        });
      });

      this.sentry = Sentry;
    }
  }

  /**
   * Returns a Winston logger configured for our lambdas.
   *
   * Note: If the lambda is executed in a `serverless-offline` context, the
   * log output to console will be pretty-printed.
   */
  getLogger() {
    const loggerFormats = [
      Winston.format.json({
        replacer: (key, value) => {
          if (value instanceof Buffer) {
            return value.toString('base64');
          }
          if (value instanceof Error) {
            return Object.fromEntries(
              Object.getOwnPropertyNames(value)
                .map((errorKey) => [errorKey, (value as any)[errorKey]]),
            );
          }
          return value;
        },
      }),
    ];

    if (this.di.isOffline) {
      loggerFormats.push(Winston.format.prettyPrint());
    }

    return Winston.createLogger({
      level: 'info',
      format: Winston.format.combine(...loggerFormats),
      transports: [new Winston.transports.Console()],
    });
  }

  /**
   * Returns the logger.
   *
   * Uses a cached Winston logger if it has been already created, otherwise it
   * creates one.
   */
  get logger() {
    if (!this.winston) {
      this.winston = this.getLogger();
    }

    return this.winston;
  }

  /**
   * Get Sentry client.
   */
  getSentry() {
    return this.sentry;
  }

  /**
   * While logging an error, we should recognise axios errors and trim down the
   * information to only what is useful for debugging.
   *
   * Keep the following keys:
   * - message.config
   * - message.message
   * - message.response?.status
   * - message.response?.data
   *
   * @param {object} error
   */
  static processAxiosError(error: AxiosError) {
    const processed: any = {
      config: error.config,
      message: error.message,
    };

    // It's pretty common for axios errors to not have a `response`,
    // for example if there was a network error or timeout.
    if (error.response) {
      processed.response = {
        status: error.response.status,
        data: error.response.data,
      };
    }

    return processed;
  }

  /**
   * Transform the original message before it is passed to the logger.
   *
   * @param message
   */
  static processMessage(message: any) {
    let processed = message;

    if (processed?.isAxiosError) {
      processed = LoggerService.processAxiosError(processed);
    }

    return processed;
  }

  /**
   * Log Error Message
   *
   * @param error object
   * @param message string
   */
  error(error: any, message = '') {
    if (sentryIsAvailable && error instanceof Error) {
      Sentry.captureException(error);
    }

    if (process.env.LUMIGO_TOKEN && error instanceof Error) {
      // todo: find out what the equivalent is in Lumigo
      // Epsagon.setError(error);
    }

    this.logger.log('error', message, { error: LoggerService.processMessage(error) });
    this.label('error', true);
    this.metric('error', 'error', true);
  }

  /**
   * Log an informational message.
   *
   * @param message
   */
  info(message: any) {
    this.logger.log('info', LoggerService.processMessage(message));
  }

  /**
   * Log an error, using `LoggerService.error` or `LoggerService.info` based
   * on `process.env.LOGGER_SOFT_WARNING`.
   *
   * Please note that `LoggerService.error` and `LoggerService.info` have
   * different signatures. The function uses the shared argument instead of
   * introducing ambiguity.
   *
   * @param error
   */
  warning(error: any) {
    const softWarningValues = ['true', '1'];

    if (softWarningValues.includes(process.env.LOGGER_SOFT_WARNING || '')) {
      return this.info(error);
    }

    return this.error(error);
  }

  /**
   * Add a label to the function's Epsagon trace.
   *
   * @param descriptor
   * @param silent If `false`, the label will also be logged. (default: false)
   */
  label(descriptor: string, silent = false) {
    if (process.env.LUMIGO_TOKEN) {
      // todo: do we need to use our `tracer` instance here?
      lumigo.addExecutionTag(descriptor, true);
    }

    if (!silent) {
      this.logger.log('info', `label - ${descriptor}`);
    }
  }

  /**
   * Add a metric to the function's Epsagon trace.
   *
   * @param descriptor
   * @param stat
   * @param silent If `false`, the metric will also be logged. (default: false)
   */
  metric(descriptor: string, stat: number | string, silent = false) {
    if (process.env.LUMIGO_TOKEN) {
      // todo: do we need to use our `tracer` instance here?
      lumigo.addExecutionTag(descriptor, stat);
    }

    if (silent === false) {
      this.logger.log('info', `metric - ${descriptor} - ${stat}`);
    }
  }

  /**
   * Log an object so that it can be inspected.
   *
   * @param action What are we doing with the object, e.g. 'Processing'
   * @param object The object to be stored in logs
   * @param level 'error', 'warning' or 'info'
   */
  object(action: string, object: any, level: 'error' | 'warning' | 'info' = 'info') {
    if (!(['error', 'warning', 'info'].includes(level))) {
      throw new Error('Unrecognised log level');
    }

    const payload = JSON.stringify(object, null, 4);

    return this[level](`${action}: '${payload}'`);
  }
}
