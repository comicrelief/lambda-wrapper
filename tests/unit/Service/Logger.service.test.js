import Winston from 'winston';

import DependencyInjection from '../../../src/DependencyInjection/DependencyInjection.class';
import LoggerService from '../../../src/Service/Logger.service';
import CONFIGURATION from '../../../src/Config/Dependencies';

const getEvent = require('../../mocks/aws/event.json');

const getContext = { invokedFunctionArn: 'my-function' };

const getLogger = (event = getEvent, context = getContext) => {
  return new LoggerService(new DependencyInjection(CONFIGURATION, event, context));
};

describe('Service/LoggerService', () => {
  const context = { invokedFunctionArn: 'my-function' };

  afterEach(() => jest.clearAllMocks());

  describe('constructor', () => {
    it('Creates a LoggerService instance', () => {
      expect(getLogger()).toBeInstanceOf(LoggerService);
    });
  });

  describe('getLogger', () => {
    it('Creates a logger instance', () => {
      const logger = getLogger(undefined, context);
      const winston = logger.getLogger();
      expect(winston.constructor.name).toEqual('DerivedLogger');
    });
  });

  describe('logger', () => {
    it('Starts as null', () => {
      const logger = getLogger(undefined, context);
      expect(logger.winston).toEqual(null);
    });

    it('Fetches a logger', () => {
      const winston = Symbol('winston');
      const logger = getLogger(undefined, context);
      jest.spyOn(Winston, 'createLogger').mockImplementation(() => winston);

      expect(logger.logger).toEqual(winston);
    });

    it("Doesn' call Winston.createLogger twice", () => {
      const winston = Symbol('winston');
      const logger = getLogger(undefined, context);
      jest.spyOn(Winston, 'createLogger').mockImplementation(() => winston);

      expect(logger.logger).toEqual(winston);
      expect(logger.logger).toEqual(winston);
      expect(logger.logger).toEqual(winston);

      expect(Winston.createLogger).toHaveBeenCalledTimes(1);
    });
  });
});
