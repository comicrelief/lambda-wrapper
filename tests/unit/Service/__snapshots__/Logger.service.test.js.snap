// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service/LoggerService error Trims down the axios error: EMPTY 1`] = `
Object {
  "config": Object {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
  "response": Object {
    "data": undefined,
    "status": undefined,
  },
}
`;

exports[`Service/LoggerService error Trims down the axios error: HTTP_417 1`] = `
Object {
  "config": Object {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
  "response": Object {
    "data": Object {
      "data": 1,
    },
    "status": 417,
  },
}
`;

exports[`Service/LoggerService error Trims down the axios error: UNDEFINED 1`] = `
Object {
  "config": Object {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
}
`;

exports[`Service/LoggerService info Trims down the axios error: EMPTY 1`] = `
Object {
  "config": Object {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
  "response": Object {
    "data": undefined,
    "status": undefined,
  },
}
`;

exports[`Service/LoggerService info Trims down the axios error: HTTP_417 1`] = `
Object {
  "config": Object {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
  "response": Object {
    "data": Object {
      "data": 1,
    },
    "status": 417,
  },
}
`;

exports[`Service/LoggerService info Trims down the axios error: UNDEFINED 1`] = `
Object {
  "config": Object {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
}
`;

exports[`Service/LoggerService object Logs a '"a string"' with level: 'error' 1`] = `"My action: '\\"a string\\"'"`;

exports[`Service/LoggerService object Logs a '"a string"' with level: 'info' 1`] = `"My action: '\\"a string\\"'"`;

exports[`Service/LoggerService object Logs a '"a string"' with level: 'warning' 1`] = `"My action: '\\"a string\\"'"`;

exports[`Service/LoggerService object Logs a '{"a":{"b":null},"c":"a string"}' with level: 'error' 1`] = `
"My action: '{
    \\"a\\": {
        \\"b\\": null
    },
    \\"c\\": \\"a string\\"
}'"
`;

exports[`Service/LoggerService object Logs a '{"a":{"b":null},"c":"a string"}' with level: 'info' 1`] = `
"My action: '{
    \\"a\\": {
        \\"b\\": null
    },
    \\"c\\": \\"a string\\"
}'"
`;

exports[`Service/LoggerService object Logs a '{"a":{"b":null},"c":"a string"}' with level: 'warning' 1`] = `
"My action: '{
    \\"a\\": {
        \\"b\\": null
    },
    \\"c\\": \\"a string\\"
}'"
`;

exports[`Service/LoggerService object Logs a '{"a":1}' with level: 'error' 1`] = `
"My action: '{
    \\"a\\": 1
}'"
`;

exports[`Service/LoggerService object Logs a '{"a":1}' with level: 'info' 1`] = `
"My action: '{
    \\"a\\": 1
}'"
`;

exports[`Service/LoggerService object Logs a '{"a":1}' with level: 'warning' 1`] = `
"My action: '{
    \\"a\\": 1
}'"
`;

exports[`Service/LoggerService object Logs a 'null' with level: 'error' 1`] = `"My action: 'null'"`;

exports[`Service/LoggerService object Logs a 'null' with level: 'info' 1`] = `"My action: 'null'"`;

exports[`Service/LoggerService object Logs a 'null' with level: 'warning' 1`] = `"My action: 'null'"`;
